### PR TITLE
Add type declarations

### PR DESCRIPTION
**Pull request by newyankeecodeshop**
I think this is good. My one concern is `MsgConstructor` because I'm not sure if it needs to support multiple arguments and whether it should have any kind of type parameterization?

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+// Definitions by: Andrew Goodale <https://github.com/newyankeecodeshop>
+// TypeScript Version: 2.3
+
+import { Component } from "react";
+
+export interface Msg {
+  tag: string;
+}
+
+export type MsgConstructor = (a?: any) => Msg;
```
Comment (hojberg) [inline index.d.ts]: `MsgConstructor` definitely takes any number of arguments of any type.

Comment (newyankeecodeshop) [reply index.d.ts]: It looks like it's only ever called with one argument:
https://github.com/hojberg/gongfu/blob/d71a61e60d7d013c36b664dba906317146dab23b/src/updater_for.ts#L6
Or maybe it's just `updaterFor` only supports one argument?

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+export type Updater = (msg: Msg) => void;
+export type Runner = (updater: Updater) => void;
+export type Setup = (onChange: Updater) => void;
+
+export interface Effect {
+  concat(that: Effect): Effect;
+
+  map(f: (a: any) => any): Effect;
```
Comment (newyankeecodeshop) [inline index.d.ts]: @hojberg Should the argument to map be of type `Msg`? It looks like it from the code, but the code uses `any` so I'm not sure.

Comment (newyankeecodeshop) [reply index.d.ts]: Or maybe it should be `map<T extends Msg>()`?

File: index.d.ts (hunk 1 above)
Comment (hojberg) [reply index.d.ts]: @newyankeecodeshop yeah `updaterFor` expects one that only has 1 argument for the nested Msg. Perhaps we should have a specific type for that.

File: index.d.ts (hunk 2 above)
Comment (hojberg) [reply index.d.ts]: @newyankeecodeshop yeah `T extends Msg`, would mean any `Msg` really. That makes sense to me.

**Review by hojberg — APPROVED** [review]

**Pull request merged**